### PR TITLE
[netcore] Implement missing bits in Mutex

### PR DESCRIPTION
**Pull request by EgorBo**
Needed for xunit console runner tool.

### CODE REVIEW

File: netcore/gen-xunit-runner/Program.cs [hunk 1]
```diff
@@ -379,7 +379,7 @@ static int Main (string[] args)
 
 				var block = ParseText<BlockSyntax> (case_template
 					.Replace ("#MSG#", msg)
-					.Replace ("#NRUN#", test.MemberDataMethod == null ? "1" : ((IEnumerable<object []>) test.MemberDataMethod.Invoke (null, null)).Count ().ToString ()));
+					.Replace ("#NRUN#", test.MemberDataMethod == null ? "1" : ((IEnumerable) test.MemberDataMethod.Invoke (null, null)).Cast<object> ().Count ().ToString ()));
```
Comment (EgorBo) [inline netcore/gen-xunit-runner/Program.cs]: Fixes `Unhandled Exception: System.InvalidCastException: Unable to cast object of type '<CopyTo_Hidden_Data>d__19' to type 'System.Collections.Generic.IEnumerable`1[System.Object[]]'.`
when it's used for System.Collections.Tests

File: mcs/class/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 2]
```diff
@@ -169,6 +169,7 @@
       <Compile Include="System.Security\DynamicSecurityMethodAttribute.cs" />
       <Compile Include="System.Threading\EventWaitHandle.cs" />
       <Compile Include="System.Threading\Interlocked.cs" />
+      <Compile Include="..\corlib\System.IO\MonoIOError.cs" />
       <Compile Include="System.Threading\Mutex.cs" />
```
Comment (vargaz) [inline mcs/class/System.Private.CoreLib/System.Private.CoreLib.csproj]: There is a ..\corlib section in the csproj, it should go there.

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs [hunk 3]
```diff
@@ -1,14 +1,80 @@
...
+			fixed (char *fixed_name = name)
+				return CreateMutex_icall (initiallyOwned, fixed_name,
+					name?.Length ?? 0, out created);
+		}
+
+		static OpenExistingResult OpenExistingWorker (string name, out Mutex result) 
+		{
+			if (name == null)
+				throw new ArgumentNullException ("name");
+
```
Comment (vargaz) [inline mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: Use nameof (name)

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs [hunk 4]
```diff
@@ -1,14 +1,80 @@
...
-		void CreateMutexCore (bool initiallyOwned, string name, out bool createdNew) => throw new NotImplementedException ();
-
-		static OpenExistingResult OpenExistingWorker (string name, out Mutex result) => throw new NotImplementedException ();
+		Mutex (IntPtr handle) => Handle = handle;
 
 		public void ReleaseMutex ()
 		{
-			throw new NotImplementedException ();
+			if (!ReleaseMutex_internal (Handle))
+				throw new ApplicationException ("Mutex is not owned");
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: This is an unlikely right exception, what does core throw?

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs [hunk 5]
```diff
@@ -1,14 +1,80 @@
...
+				throw new ArgumentNullException (nameof (name));
+
+			result = null;
+			if ((name.Length == 0) ||
+				(name.Length > 260)) {
+				return OpenExistingResult.NameInvalid;
+			}
+			
+			MonoIOError error;
+			IntPtr handle = OpenMutex_internal (name, MutexRights.Modify, out error);
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: use `out MonoIOError error` instead

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs [hunk 6]
```diff
@@ -1,14 +1,80 @@
...
+			if (handle == (IntPtr) null) {
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: prefer `handle == IntPtr.Zero`

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs [hunk 7]
```diff
@@ -1,14 +1,80 @@
...
+
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		private unsafe static extern IntPtr OpenMutex_icall (char *name, int name_length, MutexRights rights, out MonoIOError error);
+
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		private static extern bool ReleaseMutex_internal (IntPtr handle);
+	}
+
+	[Flags]
+	public enum MutexRights
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: Change the icall not to require the value

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs (hunk 5 above)
Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: These are probably the same icalls used by the normal version.

File: mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs (hunk 4 above)
Comment (filipnavara) [reply mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: Surprisingly enough that's what CoreRT throws on Unix: `throw new ApplicationException(SR.Arg_SynchronizationLockException);`

Comment (filipnavara) [reply mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: ...and CoreCLR as well: https://github.com/dotnet/coreclr/blob/cf85f56e473b6cdc4a68e9758f57dc571e14c9cb/src/System.Private.CoreLib/shared/System/Threading/Mutex.Windows.cs#L95

Comment (marek-safar) [reply mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: @stephentoub could you please confirm this^ is intentional

Comment (stephentoub) [reply mcs/class/System.Private.CoreLib/System.Threading/Mutex.cs]: It's intentional, unfortunately.  It's what netfx has used forever:
https://referencesource.microsoft.com/#mscorlib/system/threading/mutex.cs,425
and the cost/benefit to changing it at this point is not favorable.

**Pull request merged**